---
title: "scATFR: An Ensemble R package for Systematic Analysis of Transcription Factor Regulon at Single-cell Level"
output: html_notebook
---
Transcription factor regulon (TFR) 

## 1. Overview
The **scATFR** is an R package used for <u>**s**</u>ingle-<u>**c**</u>ell <u>**A**</u>ctive <u>**T**</u>ranscription <u>**F**</u>actor <u>**R**</u>egulon analysis, which combining the transcription factor regulatory information with single-cell gene expression data and evaluating the transcription factor regulon activities, identifying cell-type specific regulons and assessing the similarities between regulons and pathways.

## 2. Quick Start
Firstly, we load TF-target network from local data set and import expression data: 
```{r}
library(scATFR)
#load TF-target data 
data(dorothea_regulons)
regulons <- dfToList(df = dorothea_regulons$mm_regulons, tf_col = "tf",target_col = "target")
#load gene expression data and colData
expression_data <- readRDS(system.file("extdata", "mouse_HSC_formation_expression.rds", package = "scATFR"))
col_data <- readRDS(system.file("extdata", "mouse_HSC_formation_colData.rds", package = "scATFR"))
```
Then, we create the `atfr` object:
```{r}
atfr <- scATFR(exp_data = expression_data, col_data = col_data, regulons = regulons)
```
Next, we infer the Gene Regulatory Networks:
```{r}
atfr <- inferGRNs(x = atfr, method="sclink", ncores=6)
```
To look for GRNs
```{r}
head(GRN(atfr)[,1:5])
```
After that, we filter the gene regulons using gene regulotory network:
```{r}
atfr <- filterRegulons(x=atfr, ncores=6)
```
Now, we evaluate the regulon activity in each single cell:
```{r}
atfr <- regulonActivity(atfr, method="viper",ncores=6)
```
We can also get the cell-type specific regulons:
```{r}
#get one cell type specific
EC_sig <- findCellTypeSignature(atfr, pos_clust="EC",diff_by="stage")
#get all cell type specific
all_sig <- findAllCTSRegulons(atfr, diff_by="stage",ncores=6)
head(all_sig)
all_sig <- dplyr::tibble(all_sig)
```
We can compare the similarities between regulons and pathways:
```{r}
data(kegg_pathway)
kegg_mouse <- kegg_pathway$mouse
regulon_pathway <- inferRegulonFun(x = atfr, pathway_list = kegg_mouse, ncores=6)
```
## 3. Usage of scATFR
In this section, we introduced the detailed usage of `scATFR` step by step. 

### 3.1 Data input
The `scATFR` have two required inputs, the transcription factors with its corresponding targets and single-cell RNA sequencing expression data.

#### 3.1.1 transcription factors target regulatory data
The `scATFR` have 3 types of transcription factors target regulatory data from different resources: [DoRothEA](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-1949-z), Marbach2016 and cisTarget ()















## 3. Identifying functional TFRs
Transcription factor regulons (TFRs) are a collection of transcription factors and their targets. Here we can obtain functional TFRs by combine the TF-Target binding information with gene regulatory network(GRN).

### 1.1 Infering TFRs based on gene regulatory network(GRN) and cis-Target information
GRN identification is the first step for TFR analysis, here we integrated several common used methods for GRN analysis

#### 1.2.1 Infering GRN from gene expression data
Here we integered 5 methods to inferring GRNs from single-cell RNA sequencing data. You can using `inferGRNs()` to perform GRN identification with data `matrix` or `data.frame`. we even support servel types of objects out from common used software (eg. objects from Seurat, SingleCellExprentment, Monocle, ExpressionSet). For example:
```{r}
data(exampleTFs)
data(mouseExp)
grns <- inferGRNs(x = mouseExp, method = "sclink", use_regulator = exampleTFs, ncores=2)

```
For single-cell RNA sequencing data with time-stamped, we suggest using `SINCERITIES`. Besides, we also reorganized the scripts from `SINCERITIES` package and implement a method names `run_sincerities()` to conveniently perform GRN inference. For more usage of `run_sincerities()`, please type `?run_sincerities()`.

#### 1.2.2 Identifying TFR based on GRN and cis-regulation information

```{r}
rankings <- RcisTarget::importRankings("../ext_data/cisTarget_databases/hg38__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr.feather")
motif_anno <- data.table::fread("../ext_data/cisTarget_databases/motifs-v9-nr.hgnc-m0.001-o0.0.tbl", sep="\t")


```


### 1.2 Obtaining TFRs from DoRoThEA
Alternatively, you can also simply obtain these data from `DoRoThEA` package. The TFRs for human and mouse were deposited in latest version (1.3.2) of [dorothea](https://github.com/saezlab/dorothea) package, and can be directly extracted. Please noticed that for the standard analysis using normal tissues, `dorothea` suggest using regulons data in `dorothea_hs/dorothea_mm`, in which the tissue specific gene expression data from the GTEx portal. For abnormal tissues, the `dorothea_hs_pancancer/dorothea_mm_pancancer` is better, in which the gene expression data were  inferred from the TCGA program.

**NOTE**: only the latest version (1.3.2) include the `dorothea_hs_pancancer/dorothea_mm_pancancer` data.
```{r}
#install.packages("devtools")
#devtools::install_github("saezlab/dorothea")
library(dorothea)
#human_norm_regulons <- get(data("dorothea_hs", package = "dorothea"))
#head(human_norm_regulons)
mouse_regulons <- get(data("dorothea_mm", package = "dorothea"))
head(mouse_regulons)
human_abnorm_regulons <- get(data("dorothea_hs_pancancer", package = "dorothea"))
head(human_norm_regulons)
#mouse_abnorm_regulons <- get(data("dorothea_mm_pancancer", package = "dorothea"))
#head(mouse_abnorm_regulons)

tfrs <- tfrFromDoRoThEA(mouse_regulons, levels=c("A","B","C"))

```
Typically, there are four columns in TFRs data:

**tf**: transcription factors in gene symbols from HGNC or MGI.

**confidence**: the confidence between TFs and their Targets. Level `A` stands for highest confidence while level `E` stands for lowest confidence.

**target**: target genes in gene symbols from HGNC or MGI.

**mor**: the mode of regulation. `1` stands for activation and `-1` stands for repression.

## 2. Assessing TFR activity at single-cell level

```{r}
tfr_score <- regulonActivity(mouseExp, gene_list = tfrs, method = "viper")
```


### 2.1 VIPER

### 2.2 AUCell

### 2.3 GSVA ssGSEA

## 3. Dimensionality reduction analysis based on TFR

### 3.1 PCA tSNE and UMAP

### 3.2 Driven TFR identification

## 4. Pseudotime analysis based on TFR

## 5. Cell-type specific TFR analysis

```{r}
css_res <- calcCSS(mat = tfr_score, cell_anno = col_data, pos_clust = "T1_pre_HSC",nperm = 1000)

all_sig_css <- findAllCTSRegulons(tfr_score, cell_anno = col_data, css=0.3)

```

### 5.1 Differential TFR 

### 5.2 differential TFR based on pseudotime


## 6. Motif analysis of Cell-type specific TFR


## 7. TFR and pathway comparison analysis

```{r}
kegg_pathway <- fgsea::gmtPathways(gmt.file = "c2.cp.kegg.v7.4.symbols.gmt")

```

## Case Study

### 7. CASE1: Characterisation of cell-type specific transcription factor regulons (TFRs) during HSC formation.
In this case, we firstly performed dimension reduction and pseudotime analysis based on [Monocle3](https://cole-trapnell-lab.github.io/monocle3/) using mouse embryonic hematopoietic stem cells (HSCs) formation data ([Fan zhou, 2016](https://www.nature.com/articles/nature17997) and [Jie Zhou, 2018](https://doi.org/10.1016/j.stem.2018.11.023)). Then, we performed transcription factor regulons (TFRs) analysis such as cell-type specific TFRs identification and TFR-pathway similarity analysis using **scATFR**. Finally, we showed the visualization of the above results.

#### 7.1 Pseudotime analysis using monocle3

##### 7.1.1 Loading the data and create `cds` object
The row names of `expression_matrix` is gene symbols and the column names of `expression_matrix` is cell identifier. The `expression_colData` include three columns: sample, stage and tissue.
```{r}
library(monocle3)
#load the expression data
expression_data <- readRDS(system.file("extdata", "mouse_HSC_formation_expression.rds", package = "scATFR"))
#load the cell colData
expression_colData <- readRDS(system.file("extdata", "mouse_HSC_formation_colData.rds", package = "scATFR"))
#create cds object
cds <- new_cell_data_set(expression_data = expression_data,
                        cell_metadata = expression_colData)
```
##### 7.1.2 pre-process and dimension reduction
Then, we perform normalization and dimension reduction with object `cds`.
```{r}
#Normalize and pre-process the data
cds <- preprocess_cds(cds, num_dim = 50, method = "PCA")
#Reduce the dimensions using UMAP
cds <- reduce_dimension(cds,reduction_method = "UMAP", preprocess_method = "PCA")
#plot_cells(cds, reduction_method = "UMAP", color_cells_by="stage")
```
##### 7.1.3 Pseudotime analysis
```{r}
# Cluster the cells
cds <- cluster_cells(cds)
# Learn a graph
cds <- learn_graph(cds,use_partition = FALSE)
# Order cells
cds <- order_cells(cds)
plot_cells(cds, color_cells_by = "stage",show_trajectory_graph = TRUE)
```
Finally, to better perform regulon analysis, we need to add pseudotime and cell clusters information into colData of cds object, this is convenient for further analysis.
```{r}
pseudotime <- pseudotime(cds)
clusters <- clusters(cds)
colData(cds)$clusters <- clusters[row.names(colData(cds))]
colData(cds)$pseudotime <- pseudotime[row.names(colData(cds))]
head(colData(cds))
```
Now, we can see pseudotime information in colData of `cds`.

#### 7.2 performing regulon activity analysis using scATFR

##### 7.2.1 Loading the TF-target regulons from local data
```{r}
data(dorothea_regulons)
tf_target_tab <- dorothea_regulons$mm_regulons
mouse_regulons <- tfrFromDoRoThEA(tf_target_tab, levels=c("A","B","C","D","E"))
#add regulons into cds object
Regulon(cds) <- mouse_regulons
```
Then , we calculate the GRNs
```{r}
cds <- inferGRNs(cds, method="spearman", ncores=6)
```

```{r}
cds <- filterRegulons(x = cds, ncores=6)

cds <- regulonActivity(cds, method = "viper")

all_sig_reguon <- findAllCTSRegulons(cds, diff_by="stage", css=0.3)

```


```{r}
regulon_gene <- unique(diffRegulons(cds)$regulon)
regulon_exp <- assay(cds)[intersect(row.names(assay(cds)),regulon_gene),]

regolun_se <- SingleCellExperiment(assay=list(counts=regulon_exp))
regolun_se <- runReduceDim(regolun_se)
plotReducedDim(regolun_se,colour_by = "stage",dimred="PCA")

regolun_se <- cds[intersect(row.names(assay(cds)),regulon_gene),]
regolun_se <- runReduceDim(regolun_se)
plotReducedDim(regolun_se,dimred = "PCA",colour_by = "stage")

regolun_act <- altExp(cds)
regolun_act <- runReduceDim(regolun_act,method ="UMAP")
plotReducedDim(regolun_act,dimred = "UMAP",colour_by = "stage")

```







